crates/sui-core/src/execution_cache/object_locks.rs

```diff
@@ -0,0 +1,425 @@
+    pub(crate) async fn acquire_transaction_locks(
+        &self,
+        cache: &WritebackCache,
+        epoch_store: &AuthorityPerEpochStore,
+        owned_input_objects: &[ObjectRef],
+        transaction: VerifiedSignedTransaction,
+    ) -> SuiResult {
+        let tx_digest = *transaction.digest();
+
+        let object_ids = owned_input_objects.iter().map(|o| o.0).collect::<Vec<_>>();
+        let live_objects = Self::multi_get_objects_must_exist(cache, &object_ids)?;
+
+        // Only live objects can be locked
+        for (obj_ref, live_object) in owned_input_objects.iter().zip(live_objects.iter()) {
+            Self::verify_live_object(obj_ref, live_object)?;
+        }
+
+        let mut locks_to_write: Vec<(_, LockDetails)> =
+            Vec::with_capacity(owned_input_objects.len());
+
+        // Sort the objects before locking. This is not required by the protocol (since it's okay to
+        // reject any equivocating tx). However, this does prevent a confusing error on the client.
+        // Consider the case:
+        //   TX1: [o1, o2];
+        //   TX2: [o2, o1];
+        // If two threads race to acquire these locks, they might both acquire the first object, then
+        // error when trying to acquire the second. The error returned to the client would say that there
+        // is a conflicting tx on that object, but in fact neither object was locked and the tx was never
+        // signed. If one client then retries, they will succeed (counterintuitively).
+        let owned_input_objects = {
+            let mut o = owned_input_objects.to_vec();
+            o.sort_by_key(|o| o.0);
+            o
+        };
+
+        // Note that this function does not have to operate atomically. If there are two racing threads,
+        // then they are either trying to lock the same transaction (in which case both will succeed),
+        // or they are trying to lock the same object in two different transactions, in which case
+        // the sender has equivocated, and we are under no obligation to help them form a cert.
+        for obj_ref in owned_input_objects.iter() {
+            match self.try_set_transaction_lock(obj_ref, tx_digest, epoch_store) {
+                Ok(()) => locks_to_write.push((*obj_ref, tx_digest)),
+                Err(e) => {
+                    // revert all pending writes and return error
+                    // Note that reverting is not required for liveness, since a well formed and un-equivocating
+                    // txn cannot fail to acquire locks.
+                    // However, reverting is easy enough to do in this implementation that we do it anyway.
+                    self.clear_cached_locks(&locks_to_write);
+                    return Err(e);
+                }
+            }
+        }
+
+        // commit all writes to DB
+        epoch_store
+            .tables()?
+            .write_transaction_locks(transaction, locks_to_write.iter().cloned())?;
```

Comment:
I don't know when these errors could actually occur, but it just bothers me that we might early return without calling `clear_cached_locks`. Is it possible to turn this into an RAII so we don't need to worry about it?